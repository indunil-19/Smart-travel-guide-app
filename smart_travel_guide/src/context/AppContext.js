import React, {createContext, useReducer} from 'react';
import { initialState, reducer as AppReducer } from '../reducer/AppReducer';

export const AppContext =createContext()

export const AppProvider=({children})=>{
    const [state, dispatch]= useReducer(AppReducer, initialState)

    return(
        <AppContext.Provider  value={{state,dispatch}}>
                {children}
        </AppContext.Provider>
    )
}