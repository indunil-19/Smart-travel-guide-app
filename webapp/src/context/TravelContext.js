import React, {createContext, useReducer} from 'react';
import { initialState, Travelreducer } from '../reducer/TravelReducer';

const TravelContext =createContext()

const TravelProvider=({children})=>{
    const [state, dispatch]= useReducer(Travelreducer, initialState)

    return(
        <TravelContext.Provider  value={{state,dispatch}}>
                {children}
        </TravelContext.Provider>
    )
}

export  {TravelContext,  TravelProvider}
 