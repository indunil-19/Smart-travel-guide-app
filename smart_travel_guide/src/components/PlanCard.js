import React, { useContext, useState, useEffect } from "react";
import { StyleSheet, Alert } from "react-native";
import { Card, Title, Button, IconButton } from "react-native-paper";
import { AppContext } from "../context/AppContext";

import { Config } from "../config/config";
import { theme } from "../core/theme";
import { useNavigation } from "@react-navigation/native";
import Dialog from "react-native-dialog";
import TextInput from "./TextInput";

export const PlanCard = ({
  _id = "",
  name,
  days,
  createdDate,
  onDelete,
  travelPlan,
  displayAlert,
}) => {
  const { state, dispatch } = useContext(AppContext);
  const navigation = useNavigation();
  const [randomInt, setRandomInt] = useState();
  const [visible, setVisible] = useState(false);
  const [planName, setPlanName] = useState({ name: "", error: "" });
  const [displayName, setDisplayName] = useState(name);

  useEffect(() => {
    setRandomInt(Math.floor(Math.random() * 6));
  }, []);

  const links = [
    require("../assets/img/1.jpg"),
    require("../assets/img/2.jpg"),
    require("../assets/img/3.jpg"),
    require("../assets/img/4.jpg"),
    require("../assets/img/5.jpg"),
    require("../assets/img/6.jpg"),
  ];
  const showDialog = () => {
    if (visible) {
      setVisible(false);
    } else {
      setVisible(true);
    }
  };
  const changeName = () => {
    if (planName && planName.name.length != 0) {
      fetch(`${Config.localhost}/user/changePlanName`, {
        method: "post",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          planId: _id,
          name: planName.name,
        }),
      })
        .then((res) => res.json())
        .then((result) => {
          console.log(result);
          if (result.error) {
            dispatch({
              type: "set_notification",
              payload: {
                notification: {
                  message: result.error,
                  icon: "alert-octagon-outline",
                },
              },
            });
          } else {
            dispatch({
              type: "set_notification",
              payload: {
                notification: {
                  message: result.message,
                  icon: "check-circle-outline",
                },
              },
            });
            setDisplayName(planName.name);
          }
        })
        .catch((e) => {
          console.log(e);
        });
      console.log(planName);
      showDialog();
    } else {
      setPlanName((cur) => {
        return { name: cur.name, error: "Plan name cannot be empty." };
      });
      return;
    }
  };

  const deletePlan = () => {
    fetch(`${Config.localhost}/user/deleteTravelPlan`, {
      method: "delete",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        planId: _id,
      }),
    })
      .then((res) => res.json())
      .then((result) => {
        if (result.data) {
          onDelete(result.data);
          displayAlert();
        }

        console.log(result);
      })
      .catch((e) => {
        console.log(e);
      });
  };

  const confirmDelete = () => {
    Alert.alert(
      "Do you want to delete this plan ?",
      "Press Confirm to Delete this travel plan.",
      [
        {
          text: "Cancel",
          onPress: () => console.log("Cancel Pressed"),
          style: "cancel",
        },
        {
          text: "Confirm",
          onPress: () => {
            deletePlan();
          },
        },
      ]
    );
  };

  return (
    <Card style={styles.card} elevation={8}>
      <Card.Cover
        source={links[randomInt]}
        style={{ borderTopLeftRadius: 20, borderTopRightRadius: 20 }}
      />
      <Card.Title
        title={displayName}
        subtitle={createdDate}
        right={() => (
          <>
            <IconButton
              icon="circle-edit-outline"
              color={theme.colors.primary}
              size={25}
              onPress={() => showDialog()}
            />
            <Dialog.Container
              onBackdropPress={() => {
                showDialog();
              }}
              visible={visible}
              style={{ borderRadius: 20 }}
            >
              <Dialog.Title>Change Name of {displayName}</Dialog.Title>
              <TextInput
                placeholder="Enter new Name"
                onChangeText={(text) => {
                  setPlanName({ name: text, error: "" });
                }}
                error={planName.error}
                errorText={planName.error}
              />

              <Dialog.Button label="Cancel" color="red" onPress={showDialog} />
              <Dialog.Button
                label="Confirm"
                onPress={() => {
                  changeName();
                }}
              />
            </Dialog.Container>
          </>
        )}
      />
      <Card.Content>
        <Title>{days + " day trip"}</Title>
      </Card.Content>

      <Card.Actions
        style={{ flexDirection: "row", justifyContent: "space-between" }}
      >
        <Button
          icon="open-in-new"
          mode="contained"
          labelStyle={{ color: "white" }}
          style={{
            borderRadius: 20,
            backgroundColor: theme.colors.primary,
            marginLeft: 10,
          }}
          onPress={() => {
            dispatch({
              type: "set_travelPlan",
              payload: { travelPlan: travelPlan },
            });
            dispatch({
              type: "set_editPlan",
              payload: { editPlan: travelPlan },
            });
            dispatch({ type: "set_planId", payload: { planId: _id } });
            navigation.navigate("User Plan", {
              title: name,
              plan: travelPlan,
              id: _id,
              permissions: "editSaved",
            });
          }}
        >
          View Plan
        </Button>
        <Button
          icon="delete-outline"
          labelStyle={{ color: "white" }}
          mode="contained"
          style={{ borderRadius: 20, backgroundColor: "red", marginRight: 10 }}
          onPress={confirmDelete}
        >
          Delete
        </Button>
      </Card.Actions>
    </Card>
  );
};

const styles = StyleSheet.create({
  card: {
    width: "90%",
    alignSelf: "center",
    marginVertical: 12,
    borderRadius: 20,
  },
});
