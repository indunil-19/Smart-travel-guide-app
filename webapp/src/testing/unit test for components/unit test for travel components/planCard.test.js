import React from "react"
import renderer from "react-test-renderer";
import ReactDOM from "react-dom";
import { MemoryRouter,Router } from "react-router-dom";
import { TravelProvider } from "../../../context/TravelContext";
import "@testing-library/jest-dom/extend-expect";
import 'whatwg-fetch'
import Carousal from "../../../components/TravelPlanApp/carousal";
import { getByTestId, waitFor } from "@testing-library/dom";
import PlanCard from "../../../components/TravelPlanApp/planCard,";

describe('plan card test', () => {
    
    // it("should match snapshot",  () => {
    //     const tree = renderer.create(<MemoryRouter><TravelProvider><PlanCard/></TravelProvider></MemoryRouter>);
    //     expect(tree).toMatchSnapshot();
    // });

    it("shoud render without crashing", ()=>{
        const div = document.createElement("div");
        ReactDOM.render(<MemoryRouter><TravelProvider><PlanCard /></TravelProvider></MemoryRouter>,div)
      
       
    })
})
