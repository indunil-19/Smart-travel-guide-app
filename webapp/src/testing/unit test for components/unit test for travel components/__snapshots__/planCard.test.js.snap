// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plan card test should match snapshot 1`] = `
<div
  className="css-hx1lef"
>
  <div
    className="chakra-stack css-84zodg"
  >
    <img
      className="chakra-image__placeholder css-joy71d"
    />
    <div
      className="chakra-stack css-100f4dr"
    >
      <div
        className="chakra-stack css-84zodg"
      >
        <h3
          className="chakra-heading css-0"
        />
        <button
          aria-controls="popover-content-8940120478-2"
          aria-expanded={false}
          aria-haspopup="dialog"
          className="chakra-button css-5c2176"
          id="popover-trigger-8940120478-2"
          onClick={[Function]}
          type="button"
        >
          <svg
            className="chakra-icon css-onkibi"
            focusable={false}
            viewBox="0 0 24 24"
          >
            <g
              fill="none"
              stroke="currentColor"
              strokeLinecap="round"
              strokeWidth="2"
            >
              <path
                d="M11 4H4a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7"
              />
              <path
                d="M18.5 2.5a2.121 2.121 0 0 1 3 3L12 15l-4 1 1-4 9.5-9.5z"
              />
            </g>
          </svg>
        </button>
        <div
          className="chakra-popover__popper css-0"
          style={
            Object {
              "inset": "0 auto auto 0",
              "minWidth": "max-content",
              "position": "absolute",
              "visibility": "hidden",
            }
          }
        >
          <section
            className="chakra-popover__content css-1ab1jlb"
            id="popover-content-8940120478-2"
            onBlur={[Function]}
            onKeyDown={[Function]}
            role="dialog"
            style={
              Object {
                "--popper-arrow-shadow-color": undefined,
                "--popper-arrow-size": undefined,
                "opacity": 0,
                "transform": "scale(0.95) translateZ(0)",
                "transformOrigin": "var(--popper-transform-origin)",
                "visibility": "hidden",
              }
            }
            tabIndex={-1}
          >
            <div
              className="chakra-popover__arrow-positioner css-0"
              data-popper-arrow=""
              style={
                Object {
                  "position": "absolute",
                }
              }
            >
              <div
                className="chakra-popover__arrow css-0"
                data-popper-arrow-inner=""
              />
            </div>
            <button
              aria-label="Close"
              className="css-1w3y7qm"
              onClick={[Function]}
              type="button"
            >
              <svg
                aria-hidden={true}
                className="chakra-icon css-onkibi"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
                  fill="currentColor"
                />
              </svg>
            </button>
            <div
              className="chakra-popover__body css-1vae7t2"
              id="popover-body-8940120478-2"
            >
              <input
                className="chakra-input css-1h0nkzx"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Name"
              />
              <button
                className="chakra-button css-5c2176"
                onClick={[Function]}
                type="button"
              >
                Save
              </button>
            </div>
          </section>
        </div>
      </div>
      <h5
        className="chakra-heading css-0"
      >
         days
      </h5>
      <button
        className="chakra-button css-5c2176"
        onClick={[Function]}
        type="button"
      >
        View Plan
      </button>
      <p
        className="chakra-text css-g5hezf"
      >
        created : 
      </p>
    </div>
  </div>
</div>
`;
